### PR TITLE
Finished Exercise 2.

### PR DESCRIPTION
**Pull request by joh-lin**

### CODE REVIEW

File: Exercises/JohannesLingk/exercises_02.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# Encountering any other value when you expect an opcode indicates an error in your coding.
+        if not valid: 
+            raise Exception("invalid OPCODE")
+        #print(memory)
+# Meaning of opcodes:
+#  1 indicates addition. If you encounter the opcode 1 you should read values from two positions 
+#    of your working memory, add them, and store the result in a third position of your working memory.
+#    The three numbers immediately after your opcode indicate the memory locations to read (first two values)
+#    and write (third value) respectively. 
+#    After executing the addition you should move to the next opcode by stepping forward 4 positions.
```
Comment (GeFritz) [inline Exercises/JohannesLingk/exercises_02.py:19]: The code is functional but hard to read due to comments.

File: Exercises/JohannesLingk/exercises_02.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    result1 = [e for e in args if e.isdigit()]
+    result2 = [e for e in args if len(e) == 1]
+    return result1, result2
+
+# Test 1
+assert myFunc2("nodigit", "d", "1", "23") == (["1", "23"], ["d", "1"])
+# Test 2
+assert myFunc2("a", "b", "cd") == ([], ["a", "b"])
+# Test 3
+assert myFunc2("2345", "5.5", "12") == (["2345", "12"], [])
```
Comment (GeFritz) [inline Exercises/JohannesLingk/exercises_02.py:71]: It is up to your interpretation if something should be interpreted as a number. Nevertheless, floats like 5.5 in your tests could also be considered. I would also add a negative number.
On the other hand, your solution is nice and compact using list comprehension. Another good aspect is the usage of "asses" to test the validity of the output.

File: Exercises/JohannesLingk/exercises_02.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# Write a function that takes as input a list of integers and returns a single integer number.
+def myFunc(memory: list[int]) -> int:
+# the numbers passed as argument form the working memory of a simulated computer.
+    pointer = 0
+    while pointer < len(memory):
+        code = memory[pointer]
+# this computer will start by looking at the first value in the list passed to the function.
+# this value will contain an `opcode`. Valid opcodes are 1, 2 or 99.
+        valid = code in {1, 2, 99}
```
Comment (GeFritz) [inline Exercises/JohannesLingk/exercises_02.py:9]: adding the bool valid is not necessary but more readable.

File: Exercises/JohannesLingk/exercises_02.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if code == 1:
+            memory[memory[pointer+3]] = memory[memory[pointer+1]] + memory[memory[pointer+2]]
+            pointer += 4
+#  2 indicates multiplication. Otherwise the same rules apply as for opcode 1.
+        if code == 2:
+            memory[memory[pointer+3]] = memory[memory[pointer+1]] * memory[memory[pointer+2]]
+            pointer += 4
+# 99 indicates halt. the program should stop after encountering the opcode 99.
+        if code == 99:
+            pointer = len(memory)
```
Comment (GeFritz) [inline Exercises/JohannesLingk/exercises_02.py:29]: Using "if" statements is totally valid, match case could also work and is arguably more readable.

**Review by christianrietzsch — APPROVED** [review]
Sehr gut 
(vorallem das Arbeiten mit `assert` und Exceptions)

**Review by GeFritz — APPROVED** [review]

**Pull request merged**